File: backend/services/code_aster_intelligence.py
"""
Code_Aster Command Intelligence Module
Provides intelligent command generation, validation, and parameter mapping
for AFFE_CHAR_MECA operations in ProSolve simulation.
"""

from typing import Dict, List, Optional, Any, Tuple
from dataclasses import dataclass
from enum import Enum
import re
import json

class LoadType(Enum):
    FORCE_NODALE = "FORCE_NODALE"
    FORCE_ARETE = "FORCE_ARETE"
    FORCE_FACE = "FORCE_FACE"
    PRES_REP = "PRES_REP"
    PESANTEUR = "PESANTEUR"

class MeshTopology(Enum):
    NODE = "NODE"
    WIRE = "WIRE"
    SURFACE = "SURFACE"
    VOLUME = "VOLUME"

@dataclass
class ParameterRule:
    """Validation rule for Code_Aster parameters"""
    name: str
    required: bool = False
    type_hint: str = "float"
    min_value: Optional[float] = None
    max_value: Optional[float] = None
    depends_on: Optional[List[str]] = None
    conflicts_with: Optional[List[str]] = None
    default_value: Optional[Any] = None

@dataclass
class LoadDefinition:
    """Enhanced load definition with Code_Aster intelligence"""
    load_type: LoadType
    label: str
    description: str
    allowed_topology: List[MeshTopology]
    group_prefix: str  # GROUP_NO or GROUP_MA
    parameter_rules: List[ParameterRule]
    template_name: str
    requires_modele: bool = True
    optional_params: List[str] = None

class CodeAsterIntelligence:
    """Core intelligence engine for Code_Aster command generation"""
    
    def __init__(self):
        self.load_definitions = self._initialize_load_definitions()
        self.validation_errors = []
        self.warnings = []
    
    def _initialize_load_definitions(self) -> Dict[LoadType, LoadDefinition]:
        """Initialize enhanced load definitions with Code_Aster syntax rules"""
        
        return {
            LoadType.FORCE_NODALE: LoadDefinition(
                load_type=LoadType.FORCE_NODALE,
                label="Nodal Force",
                description="Point loads on nodes (AFFE_CHAR_MECA/FORCE_NODALE)",
                allowed_topology=[MeshTopology.NODE],
                group_prefix="GROUP_NO",
                parameter_rules=[
                    ParameterRule("FX", required=False, type_hint="float"),
                    ParameterRule("FY", required=False, type_hint="float"),
                    ParameterRule("FZ", required=False, type_hint="float"),
                    ParameterRule("MX", required=False, type_hint="float"),
                    ParameterRule("MY", required=False, type_hint="float"),
                    ParameterRule("MZ", required=False, type_hint="float"),
                ],
                template_name="force_nodale.j2"
            ),
            
            LoadType.FORCE_ARETE: LoadDefinition(
                load_type=LoadType.FORCE_ARETE,
                label="Edge Force",
                description="Linear load on edges (AFFE_CHAR_MECA/FORCE_ARETE)",
                allowed_topology=[MeshTopology.WIRE],
                group_prefix="GROUP_MA",
                parameter_rules=[
                    ParameterRule("FX", required=False, type_hint="float"),
                    ParameterRule("FY", required=False, type_hint="float"),
                    ParameterRule("FZ", required=False, type_hint="float"),
                ],
                template_name="force_arete.j2"
            ),
            
            LoadType.FORCE_FACE: LoadDefinition(
                load_type=LoadType.FORCE_FACE,
                label="Face Force",
                description="Surface traction vector (AFFE_CHAR_MECA/FORCE_FACE)",
                allowed_topology=[MeshTopology.SURFACE],
                group_prefix="GROUP_MA",
                parameter_rules=[
                    ParameterRule("FX", required=False, type_hint="float"),
                    ParameterRule("FY", required=False, type_hint="float"),
                    ParameterRule("FZ", required=False, type_hint="float"),
                ],
                template_name="force_face.j2"
            ),
            
            LoadType.PRES_REP: LoadDefinition(
                load_type=LoadType.PRES_REP,
                label="Pressure",
                description="Normal pressure (AFFE_CHAR_MECA/PRES_REP)",
                allowed_topology=[MeshTopology.SURFACE, MeshTopology.VOLUME],
                group_prefix="GROUP_MA",
                parameter_rules=[
                    ParameterRule("PRES", required=True, type_hint="float", min_value=0),
                ],
                template_name="pres_rep.j2"
            ),
            
            LoadType.PESANTEUR: LoadDefinition(
                load_type=LoadType.PESANTEUR,
                label="Gravity",
                description="Global acceleration field (AFFE_CHAR_MECA/PESANTEUR)",
                allowed_topology=[MeshTopology.VOLUME],
                group_prefix="GROUP_MA",
                parameter_rules=[
                    ParameterRule("GRAVITE", required=True, type_hint="float", min_value=0, default_value=9.81),
                    ParameterRule("DIRECTION", required=True, type_hint="vector3d", default_value=[0, 0, -1]),
                ],
                template_name="pesanteur.j2",
                optional_params=["GROUP_MA"]  # Optional for global gravity
            )
        }
    
    def validate_load_parameters(self, load_type: LoadType, parameters: Dict[str, Any]) -> Tuple[bool, List[str]]:
        """Validate parameters against Code_Aster rules"""
        self.validation_errors.clear()
        self.warnings.clear()
        
        if load_type not in self.load_definitions:
            self.validation_errors.append(f"Unknown load type: {load_type}")
            return False, self.validation_errors
        
        definition = self.load_definitions[load_type]
        
        # Check required parameters
        for rule in definition.parameter_rules:
            if rule.required and rule.name not in parameters:
                self.validation_errors.append(f"Required parameter '{rule.name}' is missing")
            elif rule.name in parameters:
                value = parameters[rule.name]
                
                # Type validation
                if rule.type_hint == "float":
                    try:
                        float_value = float(value)
                        parameters[rule.name] = float_value  # Normalize to float
                        
                        # Range validation
                        if rule.min_value is not None and float_value < rule.min_value:
                            self.validation_errors.append(f"{rule.name} must be >= {rule.min_value}")
                        if rule.max_value is not None and float_value > rule.max_value:
                            self.validation_errors.append(f"{rule.name} must be <= {rule.max_value}")
                    except (ValueError, TypeError):
                        self.validation_errors.append(f"{rule.name} must be a numeric value")
                
                elif rule.type_hint == "vector3d":
                    if isinstance(value, list) and len(value) == 3:
                        try:
                            parameters[rule.name] = [float(v) for v in value]
                        except (ValueError, TypeError):
                            self.validation_errors.append(f"{rule.name} must be a 3-element numeric vector")
                    else:
                        self.validation_errors.append(f"{rule.name} must be a 3-element vector")
        
        # Check for at least one force component for force loads
        if load_type in [LoadType.FORCE_NODALE, LoadType.FORCE_ARETE, LoadType.FORCE_FACE]:
            has_force = any(param in parameters for param in ["FX", "FY", "FZ"])
            if not has_force:
                self.warnings.append("No force components specified. Load will have no effect.")
        
        return len(self.validation_errors) == 0, self.validation_errors
    
    def generate_command_syntax(self, load_type: LoadType, parameters: Dict[str, Any], 
                               target_group: str = "", result_name: str = "load_1") -> Dict[str, Any]:
        """Generate complete AFFE_CHAR_MECA command syntax"""
        
        # Validate first
        is_valid, errors = self.validate_load_parameters(load_type, parameters)
        if not is_valid:
            return {
                "status": "error",
                "errors": errors,
                "command": None
            }
        
        definition = self.load_definitions[load_type]
        
        # Build command structure
        command = {
            "status": "success",
            "load_type": load_type.value,
            "result_name": result_name,
            "template": definition.template_name,
            "parameters": {}
        }
        
        # Add MODELE if required
        if definition.requires_modele:
            command["parameters"]["MODELE"] = "{{ modele }}"  # Jinja2 template variable
        
        # Add group specification
        if target_group or load_type != LoadType.PESANTEUR:
            group_key = definition.group_prefix
            command["parameters"][group_key] = f"'{target_group}'"
        
        # Add load-specific parameters
        load_block = {}
        for rule in definition.parameter_rules:
            if rule.name in parameters:
                value = parameters[rule.name]
                if rule.type_hint == "vector3d":
                    load_block[rule.name] = f"({value[0]}, {value[1]}, {value[2]})"
                else:
                    load_block[rule.name] = value
        
        # Add the load block
        command["parameters"][load_type.value] = f"_F({', '.join([f'{k}={v}' for k, v in load_block.items()])})"
        
        # Add optional AFFE_CHAR_MECA parameters
        optional_params = {
            "DOUBLE_LAGRANGE": "'NON'",
            "INFO": 1,
            "VERI_NORM": "'NON'",
            "VERI_AFFE": "'NON'"
        }
        command["parameters"].update(optional_params)
        
        return command
    
    def get_frontend_metadata(self, load_type: LoadType) -> Dict[str, Any]:
        """Get metadata for frontend LoadConfig component"""
        if load_type not in self.load_definitions:
            return {}
        
        definition = self.load_definitions[load_type]
        
        return {
            "load_type": load_type.value,
            "label": definition.label,
            "description": definition.description,
            "allowed_topology": [t.value for t in definition.allowed_topology],
            "group_prefix": definition.group_prefix,
            "parameter_rules": [
                {
                    "name": rule.name,
                    "required": rule.required,
                    "type_hint": rule.type_hint,
                    "min_value": rule.min_value,
                    "max_value": rule.max_value,
                    "default_value": rule.default_value
                }
                for rule in definition.parameter_rules
            ],
            "validation_hints": self._get_validation_hints(load_type)
        }
    
    def _get_validation_hints(self, load_type: LoadType) -> List[str]:
        """Get validation hints for frontend"""
        hints = []
        
        if load_type == LoadType.PRES_REP:
            hints.append("Pressure must be positive (Pa)")
        elif load_type == LoadType.PESANTEUR:
            hints.append("Gravity typically 9.81 m/s²")
            hints.append("Direction: (0,0,-1) for standard Earth gravity")
        elif load_type in [LoadType.FORCE_NODALE, LoadType.FORCE_ARETE, LoadType.FORCE_FACE]:
            hints.append("At least one force component required")
            if load_type == LoadType.FORCE_NODALE:
                hints.append("Moments only available for nodal forces")
        
        return hints

# Singleton instance
code_aster_intelligence = CodeAsterIntelligence()

File: frontend/src/lib/codeAsterIntelligence.ts
/**
 * Code_Aster Intelligence Engine - Frontend Module
 * Provides intelligent command generation, validation, and parameter mapping
 * for AFFE_CHAR_MECA operations directly in the frontend
 */

// --- Core Types and Enums ---

export const LoadType = {
    FORCE_NODALE: 'FORCE_NODALE',
    FORCE_ARETE: 'FORCE_ARETE',
    FORCE_FACE: 'FORCE_FACE',
    PRES_REP: 'PRES_REP',
    PESANTEUR: 'PESANTEUR'
} as const

export type LoadType = typeof LoadType[keyof typeof LoadType]

export const MeshTopology = {
    NODE: 'NODE',
    WIRE: 'WIRE',
    SURFACE: 'SURFACE',
    VOLUME: 'VOLUME'
} as const

export type MeshTopology = typeof MeshTopology[keyof typeof MeshTopology]

export interface ParameterRule {
    name: string
    required: boolean
    typeHint: 'float' | 'vector3d' | 'string'
    minValue?: number
    maxValue?: number
    dependsOn?: string[]
    conflictsWith?: string[]
    defaultValue?: any
    description?: string
    unit?: string
}

export interface LoadDefinition {
    loadType: LoadType
    label: string
    description: string
    allowedTopology: MeshTopology[]
    groupPrefix: 'GROUP_NO' | 'GROUP_MA'
    parameterRules: ParameterRule[]
    requiresModele: boolean
    optionalParams: string[]
    validationHints: string[]
    examples: string[]
}

export interface ValidationResult {
    isValid: boolean
    errors: string[]
    warnings: string[]
}

export interface CommandStructure {
    status: 'success' | 'error'
    loadType: string
    resultName: string
    command?: string
    parameters?: Record<string, any>
    errors?: string[]
}

export interface LoadParameters {
    [key: string]: number | number[] | string
}

// --- Code_Aster Intelligence Class ---

export class CodeAsterIntelligence {
    private static instance: CodeAsterIntelligence
    private loadDefinitions: Map<LoadType, LoadDefinition>

    private constructor() {
        this.loadDefinitions = this.initializeLoadDefinitions()
    }

    public static getInstance(): CodeAsterIntelligence {
        if (!CodeAsterIntelligence.instance) {
            CodeAsterIntelligence.instance = new CodeAsterIntelligence()
        }
        return CodeAsterIntelligence.instance
    }

    private initializeLoadDefinitions(): Map<LoadType, LoadDefinition> {
        const definitions = new Map<LoadType, LoadDefinition>()

        // FORCE_NODALE Definition
        definitions.set(LoadType.FORCE_NODALE, {
            loadType: LoadType.FORCE_NODALE,
            label: 'Nodal Force',
            description: 'Point loads on nodes (AFFE_CHAR_MECA/FORCE_NODALE)',
            allowedTopology: [MeshTopology.NODE],
            groupPrefix: 'GROUP_NO',
            parameterRules: [
                {
                    name: 'FX',
                    required: false,
                    typeHint: 'float',
                    description: 'Force component in X direction',
                    unit: 'N'
                },
                {
                    name: 'FY',
                    required: false,
                    typeHint: 'float',
                    description: 'Force component in Y direction',
                    unit: 'N'
                },
                {
                    name: 'FZ',
                    required: false,
                    typeHint: 'float',
                    description: 'Force component in Z direction',
                    unit: 'N'
                },
                {
                    name: 'MX',
                    required: false,
                    typeHint: 'float',
                    description: 'Moment component about X axis',
                    unit: 'N.m'
                },
                {
                    name: 'MY',
                    required: false,
                    typeHint: 'float',
                    description: 'Moment component about Y axis',
                    unit: 'N.m'
                },
                {
                    name: 'MZ',
                    required: false,
                    typeHint: 'float',
                    description: 'Moment component about Z axis',
                    unit: 'N.m'
                }
            ],
            requiresModele: true,
            optionalParams: ['DOUBLE_LAGRANGE', 'INFO', 'VERI_NORM', 'VERI_AFFE'],
            validationHints: [
                'At least one force component (FX, FY, FZ) is recommended',
                'Moments (MX, MY, MZ) only available for nodal forces',
                'Forces are applied at node locations'
            ],
            examples: [
                'FX = 1000.0',
                'FY = -500.0, FZ = 200.0',
                'FX = 1000.0, MY = 50.0'
            ]
        })

        // FORCE_ARETE Definition  
        definitions.set(LoadType.FORCE_ARETE, {
            loadType: LoadType.FORCE_ARETE,
            label: 'Edge Force',
            description: 'Linear load on edges (AFFE_CHAR_MECA/FORCE_ARETE)',
            allowedTopology: [MeshTopology.WIRE],
            groupPrefix: 'GROUP_MA',
            parameterRules: [
                {
                    name: 'FX',
                    required: false,
                    typeHint: 'float',
                    description: 'Linear force density in X direction',
                    unit: 'N/m'
                },
                {
                    name: 'FY',
                    required: false,
                    typeHint: 'float',
                    description: 'Linear force density in Y direction',
                    unit: 'N/m'
                },
                {
                    name: 'FZ',
                    required: false,
                    typeHint: 'float',
                    description: 'Linear force density in Z direction',
                    unit: 'N/m'
                },
                {
                    name: 'MX',
                    required: false,
                    typeHint: 'float',
                    description: 'Linear moment density about X axis',
                    unit: 'N·m/m'
                },
                {
                    name: 'MY',
                    required: false,
                    typeHint: 'float',
                    description: 'Linear moment density about Y axis',
                    unit: 'N·m/m'
                },
                {
                    name: 'MZ',
                    required: false,
                    typeHint: 'float',
                    description: 'Linear moment density about Z axis',
                    unit: 'N·m/m'
                }
            ],
            requiresModele: true,
            optionalParams: ['DOUBLE_LAGRANGE', 'INFO', 'VERI_NORM', 'VERI_AFFE'],
            validationHints: [
                'Forces are distributed per unit length (N/m)',
                'At least one force component recommended',
                'Applied to wire/edge elements'
            ],
            examples: [
                'FX = 100.0',
                'FY = -50.0, FZ = 25.0'
            ]
        })

        // FORCE_FACE Definition
        definitions.set(LoadType.FORCE_FACE, {
            loadType: LoadType.FORCE_FACE,
            label: 'Face Force',
            description: 'Surface traction vector (AFFE_CHAR_MECA/FORCE_FACE)',
            allowedTopology: [MeshTopology.SURFACE],
            groupPrefix: 'GROUP_MA',
            parameterRules: [
                {
                    name: 'FX',
                    required: false,
                    typeHint: 'float',
                    description: 'Surface traction in X direction',
                    unit: 'N/m²'
                },
                {
                    name: 'FY',
                    required: false,
                    typeHint: 'float',
                    description: 'Surface traction in Y direction',
                    unit: 'N/m²'
                },
                {
                    name: 'FZ',
                    required: false,
                    typeHint: 'float',
                    description: 'Surface traction in Z direction',
                    unit: 'N/m²'
                }
            ],
            requiresModele: true,
            optionalParams: ['DOUBLE_LAGRANGE', 'INFO', 'VERI_NORM', 'VERI_AFFE'],
            validationHints: [
                'Forces are distributed per unit area (N/m²)',
                'At least one force component recommended',
                'Applied to surface/face elements'
            ],
            examples: [
                'FZ = -1000.0',
                'FX = 500.0, FY = -300.0'
            ]
        })

        // PRES_REP Definition
        definitions.set(LoadType.PRES_REP, {
            loadType: LoadType.PRES_REP,
            label: 'Pressure',
            description: 'Normal pressure (AFFE_CHAR_MECA/PRES_REP)',
            allowedTopology: [MeshTopology.SURFACE, MeshTopology.VOLUME],
            groupPrefix: 'GROUP_MA',
            parameterRules: [
                {
                    name: 'PRES',
                    required: true,
                    typeHint: 'float',
                    minValue: 0,
                    description: 'Normal pressure magnitude',
                    unit: 'Pa',
                    defaultValue: 0
                }
            ],
            requiresModele: true,
            optionalParams: ['DOUBLE_LAGRANGE', 'INFO', 'VERI_NORM', 'VERI_AFFE'],
            validationHints: [
                'Pressure must be positive (Pa)',
                'Applied normal to surface (positive = into surface)',
                'Can be applied to surface or volume elements'
            ],
            examples: [
                'PRES = 101325.0',
                'PRES = 50000.0'
            ]
        })

        // PESANTEUR Definition
        definitions.set(LoadType.PESANTEUR, {
            loadType: LoadType.PESANTEUR,
            label: 'Gravity',
            description: 'Global acceleration field (AFFE_CHAR_MECA/PESANTEUR)',
            allowedTopology: [MeshTopology.VOLUME],
            groupPrefix: 'GROUP_MA',
            parameterRules: [
                {
                    name: 'GRAVITE',
                    required: true,
                    typeHint: 'float',
                    minValue: 0,
                    description: 'Gravitational acceleration magnitude',
                    unit: 'm/s²',
                    defaultValue: 9.81
                },
                {
                    name: 'DIRECTION',
                    required: true,
                    typeHint: 'vector3d',
                    description: 'Gravity direction vector',
                    defaultValue: [0, 0, -1]
                }
            ],
            requiresModele: true,
            optionalParams: ['GROUP_MA', 'DOUBLE_LAGRANGE', 'INFO', 'VERI_NORM', 'VERI_AFFE'],
            validationHints: [
                'Standard Earth gravity: 9.81 m/s²',
                'Direction (0,0,-1) for standard downward gravity',
                'GROUP_MA optional - if omitted, applies to entire model'
            ],
            examples: [
                'GRAVITE = 9.81, DIRECTION = (0, 0, -1)',
                'GRAVITE = 1.62, DIRECTION = (0, 0, -1)'
            ]
        })

        return definitions
    }

    // --- Public API Methods ---

    public validateLoadParameters(loadType: LoadType, parameters: LoadParameters): ValidationResult {
        const errors: string[] = []
        const warnings: string[] = []

        const definition = this.loadDefinitions.get(loadType)
        if (!definition) {
            errors.push(`Unknown load type: ${loadType}`)
            return { isValid: false, errors, warnings }
        }

        // Check required parameters
        for (const rule of definition.parameterRules) {
            if (rule.required && !(rule.name in parameters)) {
                errors.push(`Required parameter '${rule.name}' is missing`)
            } else if (rule.name in parameters) {
                const value = parameters[rule.name]

                // Type validation
                if (rule.typeHint === 'float') {
                    const numValue = Number(value)
                    if (isNaN(numValue)) {
                        errors.push(`${rule.name} must be a numeric value`)
                    } else {
                        parameters[rule.name] = numValue // Normalize to number

                        // Range validation
                        if (rule.minValue !== undefined && numValue < rule.minValue) {
                            errors.push(`${rule.name} must be >= ${rule.minValue}`)
                        }
                        if (rule.maxValue !== undefined && numValue > rule.maxValue) {
                            errors.push(`${rule.name} must be <= ${rule.maxValue}`)
                        }
                    }
                } else if (rule.typeHint === 'vector3d') {
                    if (Array.isArray(value) && value.length === 3) {
                        const validVector = value.every(v => !isNaN(Number(v)))
                        if (validVector) {
                            parameters[rule.name] = value.map(v => Number(v))
                        } else {
                            errors.push(`${rule.name} must be a 3-element numeric vector`)
                        }
                    } else {
                        errors.push(`${rule.name} must be a 3-element vector`)
                    }
                }
            }
        }

        // Load-specific validation
        if (loadType === LoadType.FORCE_NODALE || loadType === LoadType.FORCE_ARETE || loadType === LoadType.FORCE_FACE) {
            const hasForce = ['FX', 'FY', 'FZ'].some(param => param in parameters && parameters[param] !== 0)
            const hasMoment = ['MX', 'MY', 'MZ'].some(param => param in parameters && parameters[param] !== 0)

            if (!hasForce && !hasMoment) {
                warnings.push('No force or moment components specified. Load will have no effect.')
            }
        }

        return {
            isValid: errors.length === 0,
            errors,
            warnings
        }
    }

    public generateCommandSyntax(
        loadType: LoadType,
        parameters: LoadParameters,
        targetGroup: string = '',
        resultName: string = 'load_1'
    ): CommandStructure {
        // Validate first
        const validation = this.validateLoadParameters(loadType, parameters)
        if (!validation.isValid) {
            return {
                status: 'error',
                loadType: loadType,
                resultName,
                errors: validation.errors
            }
        }

        const definition = this.loadDefinitions.get(loadType)!
        const args: string[] = []

        // Add MODELE (always required for AFFE_CHAR_MECA)
        args.push('MODELE = MODELE')

        // Add group specification
        if (targetGroup || loadType !== LoadType.PESANTEUR) {
            const groupKey = definition.groupPrefix
            args.push(`${groupKey} = '${targetGroup}'`)
        }

        // Add load-specific parameters
        const loadParams: string[] = []

        for (const rule of definition.parameterRules) {
            if (rule.name in parameters) {
                const value = parameters[rule.name]
                if (rule.typeHint === 'vector3d') {
                    const vectorValue = value as number[]
                    loadParams.push(`${rule.name} = (${vectorValue.join(', ')})`)
                } else {
                    loadParams.push(`${rule.name} = ${value}`)
                }
            }
        }

        // Add the load block
        if (loadParams.length > 0) {
            args.push(`${loadType} = _F(\n    ${loadParams.join(',\n    ')}\n)`)
        }

        // Add optional AFFE_CHAR_MECA parameters from the parameters object
        const optionalParams = {
            'DOUBLE_LAGRANGE': parameters.DOUBLE_LAGRANGE || '\'NON\'',
            'INFO': parameters.INFO !== undefined ? parameters.INFO : 1,
            'VERI_NORM': parameters.VERI_NORM || '\'NON\'',
            'VERI_AFFE': parameters.VERI_AFFE || '\'NON\''
        }

        // Only add optional parameters that are different from defaults or explicitly set
        if (optionalParams.DOUBLE_LAGRANGE !== '\'NON\'') {
            args.push(`DOUBLE_LAGRANGE = ${optionalParams.DOUBLE_LAGRANGE}`)
        }
        if (optionalParams.INFO !== 1) {
            args.push(`INFO = ${optionalParams.INFO}`)
        }
        if (optionalParams.VERI_NORM !== '\'NON\'') {
            args.push(`VERI_NORM = ${optionalParams.VERI_NORM}`)
        }
        if (optionalParams.VERI_AFFE !== '\'NON\'') {
            args.push(`VERI_AFFE = ${optionalParams.VERI_AFFE}`)
        }

        // Generate final command
        const command = `${resultName} = AFFE_CHAR_MECA(\n    ${args.join(',\n    ')}\n);`

        return {
            status: 'success',
            loadType: loadType,
            resultName,
            command,
            parameters: {
                ...parameters,
                targetGroup,
                resultName
            }
        }
    }

    public getLoadDefinition(loadType: LoadType): LoadDefinition | undefined {
        return this.loadDefinitions.get(loadType)
    }

    public getAllLoadDefinitions(): LoadDefinition[] {
        return Array.from(this.loadDefinitions.values())
    }

    public getValidationHints(loadType: LoadType): string[] {
        const definition = this.loadDefinitions.get(loadType)
        return definition?.validationHints || []
    }

    public getSyntaxHelp(): Record<string, any> {
        return {
            affe_char_meca: {
                description: 'Assign loads and boundary conditions to mechanical model',
                syntax: 'char_meca = AFFE_CHAR_MECA(MODELE = mo, ...)',
                common_parameters: {
                    'MODELE': 'Mechanical model (required)',
                    'DOUBLE_LAGRANGE': 'Double Lagrange method (\'OUI\'/\'NON\')',
                    'INFO': 'Information level (1/2)',
                    'VERI_NORM': 'Normal verification (\'OUI\'/\'NO\')',
                    'VERI_AFFE': 'Assignment verification (\'OUI\'/\'NO\')'
                },
                load_types: {
                    'FORCE_NODALE': 'Point loads on nodes',
                    'FORCE_ARETE': 'Linear loads on edges',
                    'FORCE_FACE': 'Surface traction on faces',
                    'PRES_REP': 'Normal pressure',
                    'PESANTEUR': 'Gravity field'
                }
            }
        }
    }
}

// Export singleton instance
export const codeAsterIntelligence = CodeAsterIntelligence.getInstance()
